File: backend/app/repo_ref.py
"""Parsing for GitHub repo references.

Lives apart from ``app.github`` so the request schemas can normalize a repo on
the way in — ``app.github`` imports ``app.schemas``, so it cannot be imported
from there without a cycle.
"""

from __future__ import annotations

import re

_REPO_RE = re.compile(r"^[\w.-]+/[\w.-]+$")
_URL_RE = re.compile(r"github\.com[/:]([\w.-]+/[\w.-]+?)(?:\.git)?/?$")


class InvalidRepoRef(ValueError):
    """The given string is not a usable ``owner/name`` reference."""


def normalize_repo(repo: str) -> str:
    """Accept 'owner/name' or a full GitHub URL and return 'owner/name'.

    Stored values must be normalized: the UI interpolates them straight into
    ``https://github.com/{repo}`` hrefs, so a full clone URL doubles up.
    """
    repo = repo.strip()
    match = _URL_RE.search(repo)
    if match:
        repo = match.group(1)
    if not _REPO_RE.match(repo):
        raise InvalidRepoRef(f"Invalid repo reference: {repo!r}")
    return repo

File: android/core/src/main/kotlin/net/nickknows/ideabrd/core/RepoRef.kt
package net.nickknows.ideabrd.core

/**
 * Parsing for GitHub repo references — the port of `backend/app/repo_ref.py`.
 *
 * Stored values must be normalized to "owner/name": the UI interpolates them
 * straight into `https://github.com/{repo}` links, so a full clone URL doubles
 * up.
 */

private val REPO_RE = Regex("""^[\w.-]+/[\w.-]+$""")
private val URL_RE = Regex("""github\.com[/:]([\w.-]+/[\w.-]+?)(?:\.git)?/?$""")

class InvalidRepoRef(message: String) : IllegalArgumentException(message)

/** Accept "owner/name" or a full GitHub URL and return "owner/name". */
fun normalizeRepo(repo: String): String {
    var value = repo.trim()
    URL_RE.find(value)?.let { value = it.groupValues[1] }
    if (!REPO_RE.matches(value)) throw InvalidRepoRef("Invalid repo reference: $repo")
    return value
}

/** The same, for the parser, where an unusable value costs the field not the file. */
fun normalizeRepoOrNull(repo: String): String? =
    try {
        normalizeRepo(repo)
    } catch (_: InvalidRepoRef) {
        null
    }
